### PR TITLE
 Dialer.RequireStartTLS to configure STARTTLS security

### PR DESCRIPTION
**Pull request by sfllaw**
It is possible to use a man-in-the-middle attack to strip STARTTLS security from an SMTP connection. Because all modern SMTP servers support STARTTLS, we recommend that `Dialer.RequireStartTLS` be set to `mail.MandatoryStartTLS`.

For backwards compatibility, this library defaults to `mail.OpportunisticStartTLS`, which means that the client will fallback to a cleartext connection if the SMTP server does not support the STARTTLS extension.

This has no effect on SMTPS connections where TLS wraps the connection itself.

Closes go-mail/mail#23.

### CODE REVIEW

**Comment by sfllaw** [comment]
@ivy Please review this draft patch. I'd be happy to take any suggestions on how to make things simpler when it comes to setting options on the `Dialer`.

File: smtp.go [hunk 1]
```diff
@@ -27,9 +27,17 @@ type Dialer struct {
 	// most cases since the authentication mechanism should use the STARTTLS
 	// extension instead.
 	SSL bool
-	// TSLConfig represents the TLS configuration used for the TLS (when the
+	// TLSConfig represents the TLS configuration used for the TLS (when the
```
Comment (ivy) [inline smtp.go:30]: Good catch!

File: smtp.go [hunk 2]
```diff
@@ -27,9 +27,17 @@ type Dialer struct {
...
 	// STARTTLS extension is used) or SSL connection.
 	TLSConfig *tls.Config
+	// RequireStartTLS represents the TLS security level required to
+	// communicate with the SMTP server.
+	//
+	// This defaults to OpportunisticStartTLS for backwards compatibility,
+	// but we recommend MandatoryStartTLS for all modern SMTP servers.
+	//
+	// This option has no effect if SSL is set to true.
+	RequireStartTLS StartTLSSecurity
```
Comment (ivy) [inline smtp.go]: I think `StartTLSPolicy` would be a more fitting name for both `RequireStartTLS` and `StartTLSSecurity`.

File: smtp.go [hunk 3]
```diff
@@ -95,8 +103,16 @@ func (d *Dialer) Dial() (SendCloser, error) {
 		}
 	}
 
-	if !d.SSL {
-		if ok, _ := c.Extension("STARTTLS"); ok {
+	if !d.SSL && d.RequireStartTLS != NoStartTLS {
+		ok, _ := c.Extension("STARTTLS")
+		if !ok && d.RequireStartTLS == MandatoryStartTLS {
+			err := fmt.Errorf(
```
Comment (ivy) [inline smtp.go]: We should export an error for this (e.g. `ErrStartTLSUnsupported`).

**Review by ivy — CHANGES_REQUESTED** [review]
I only have a couple nitpicks. Everything else looks good to me!

File: smtp.go (hunk 2 above)
Comment (sfllaw) [reply smtp.go]: Fixed in 0c066ca.

File: smtp.go (hunk 3 above)
Comment (sfllaw) [reply smtp.go]: Fixed in d3eed43, but named `StartTLSUnsupportedError` to follow Go conventions: https://blog.golang.org/error-handling-and-go

Comment (ivy) [reply smtp.go]: That much better than my suggestion. I like your use of `StartTLSPolicy.String()` for handling the error message.😄

**Comment by ivy** [comment]
@sfllaw Thanks for implementing this along with the changes I suggested. :+1:

**Comment by sfllaw** [comment]
@ivy Thanks for catching the missing change in `CHANGELOG.md`.

Do you have plans to release 2.2.0 soon? I'd like to tighten up STARTTLS in some of our code. :smile:

**Comment by ivy** [comment]
@sfllaw Sure thing, I tagged and released 2.2.0. Thanks again for helping tighten up security! :+1:

**Pull request closed without merging**